### ISSUE TITLE
[Feature] Support GLM-5.2 MXFP4 with DSA/NSA and Mega-MoE

### ISSUE THREAD

**Issue by LinearPro** [sgl-project/sglang#36447]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

## Motivation

We would like to upstream support for GLM-5.2-MXFP4 in SGLang.

The model uses the `GlmMoeDsaForCausalLM` architecture and a mixed-precision
checkpoint:

- Dense attention and shared-expert weights use FP8 E4M3 with block scales.
- Routed experts use packed FP4 E2M1 weights with UE8M0 scales.
- Embedding, normalization, router and lm_head parameters remain BF16/F32.
- The model uses a DSA/NSA indexer with `topk=2048`.
- The maximum context length is up to 1M tokens.
- The intended deployment uses 8-way tensor/expert parallelism, Mega-MoE,
  NSA sparse attention, context parallelism and EAGLE/MTP.

## Prototype implementation

Our prototype contains the following changes:

1. GLM DSA model mapping and mixed FP8/FP4 expert weight loading.
2. Packed FP4 expert layout handling and UE8M0 scale conversion.
3. Mega-MoE symmetric-buffer dispatch with DeepGEMM FP8/FP4 expert
   computation.
4. Fused activation quantization and MXFP4 scale layout conversion.
5. NSA paged K/S gather and memory-budgeted MQA logits chunking.
6. A deterministic CUDA TopK implementation specialized for `topk=2048`.
7. NSA context-parallel and MTP top-k metadata propagation.
8. CUDA JIT and Triton optimizations for the relevant model path.

The prototype distinguishes the MoE runner from the A2A backend:
`flashinfer_mxfp4` selects the MoE runner interface, while `megamoe` selects
the Mega-MoE A2A path.

## Intended launch configuration

```bash
export SGLANG_DSV4_FP4_EXPERTS=1
export SGLANG_GLM51_MXFP4_MODE=1
export SGLANG_OPT_USE_DEEPGEMM_MEGA_MOE=1
export SGLANG_OPT_FIX_MEGA_MOE_MEMORY=1
export SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_FP4_ACTS=1
export SGLANG_OPT_DEEPGEMM_MEGA_MOE_USE_MXF4_KIND=1
export SGLANG_NSA_TOPK_IMPL=cuda_stable

python -m sglang.launch_server \
  --model-path <MODEL_PATH> \
  --tp-size 8 \
  --moe-runner-backend flashinfer_mxfp4 \
  --moe-a2a-backend megamoe \
  --attention-backend nsa \
  --enable-nsa-prefill-context-parallel \
  --nsa-prefill-backend flashmla_sparse \
  --disable-cuda-graph \
  --disable-overlap-schedule

### Related resources

```markdown
- Model architecture: `GlmMoeDsaForCausalLM`
- Relevant SGLang areas:
  - `python/sglang/srt/layers/quantization`
  - `python/sglang/srt/layers/moe`
  - `python/sglang/srt/layers/attention/nsa`
  - `python/sglang/python/sglang/jit_kernel`
- The prototype includes a detailed source-level analysis and can be
  provided as a draft PR after the upstream design is confirmed.

**Comment by LinearPro** [comment]
A prototype implementation is available locally. We are preparing a draft
PR based on the latest `upstream/main` and will add reproducible accuracy and
performance results once the target GPU environment is available.
